### DISCUSSION TITLE
Warnings not captured in async test

### DISCUSSION THREAD

**Discussion by blokje** [pytest-dev/pytest#8533] [Q&A]
Hi All, 

I have the following code and I'm curious about what I'm doing wrong. It seems that for async functions the warnings are not captured.

The code is as following
```python
# content of test_show_warnings.py
import warnings


def api_v1_for_sync():
    warnings.warn(UserWarning("[sync] api v1, should use functions from v2"))
    return 1

def api_v1_for_async():
    warnings.warn(UserWarning("[async] api v1, should use functions from v2"))
    return 1


def test_sync():
    assert api_v1_for_sync() == 1

async def test_async():
    assert api_v1_for_async() == 1
```

And when running the test I would assume that both warnings are captured. Unfortunately only the sync warnings shows.
```
========================================= test session starts ==========================================
platform darwin -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- /Users/rick/.local/share/virtualenvs/mock-qxkUTm3B/bin/python
cachedir: .pytest_cache
rootdir: /private/tmp/mock
plugins: cov-2.11.1, aiohttp-0.3.0, mock-3.5.1
collected 2 items                                                                                      

test_warning.py::test_sync PASSED                                                                [ 50%]
test_warning.py::test_async PASSED                                                               [100%]

=========================================== warnings summary ===========================================
test_warning.py::test_sync
  /private/tmp/mock/test_warning.py:6: UserWarning: [sync] api v1, should use functions from v2
    warnings.warn(UserWarning("[sync] api v1, should use functions from v2"))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
===================================== 2 passed, 1 warning in 0.02s =====================================
```

When I change the async test to run synchrounous both warnings show up.

```
=========================================== warnings summary ===========================================
test_warning.py::test_sync
  /private/tmp/mock/test_warning.py:6: UserWarning: [sync] api v1, should use functions from v2
    warnings.warn(UserWarning("[sync] api v1, should use functions from v2"))

test_warning.py::test_async
  /private/tmp/mock/test_warning.py:10: UserWarning: [async] api v1, should use functions from v2
    warnings.warn(UserWarning("[async] api v1, should use functions from v2"))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
==================================== 2 passed, 2 warnings in 0.01s =====================================
```

Is this expected behaviour, is async not (well) supported for pytest or am I missing something.

**Comment by portyu9** [comment]
Hi @blokje 

The first thing to verify is that the coroutine is actually being executed.

pytest core does not itself provide an asyncio event-loop runner for arbitrary `async def` tests; async tests need an appropriate plugin such as `pytest-asyncio` or AnyIO integration.

For example with pytest-asyncio:

```python
@pytest.mark.asyncio
async def test_something():
    api_v1_for_async()
```

Warnings emitted while that body genuinely executes are handled by pytest's normal warning capture.

Given the versions in the post (`pytest 6.2.3` and old async/aiohttp plugin versions), a plugin accepting the coroutine without executing it correctly is a much more plausible explanation than pytest's warning subsystem selectively ignoring warnings.

A simple diagnostic is:

```python
@pytest.mark.asyncio
async def test_something():
    assert False
```

If that does not fail, the coroutine body is not being run by the async plugin/configuration you expect.

I would reproduce this on current pytest plus a current async plugin before investigating warning capture itself.

Please mark this answer as accepted if it helped, thank you.
